### ISSUE TITLE
Tiled rendering has no support for isometric/staggered orientations

### ISSUE THREAD

**Issue by Vanit** [prime31/Nez#124]
Was hoping to lean on Nez for Tiled rendering, but quickly learnt that only orthogonal is supported.

Expected:
![tiledexpected](https://cloud.githubusercontent.com/assets/852857/19270407/3d2e24fe-9003-11e6-960e-1a35265574b7.PNG)

Actual:
![tiledactual](https://cloud.githubusercontent.com/assets/852857/19270410/424a573c-9003-11e6-8cc3-8f27a60ea1d7.PNG)

Took a cursory read of the Tiled classes and I can already see it'd be no small undertaking to get it supported. I'm not sure what the best solution is, as it would probably require a significant restructure to handle this nicely. Currently the orthogonal orientation is just assumed right through the Tiled hierarchy.

**Comment by prime31** [comment]
As of now I have no plans to add orthogonal myself. Nez generally gets features as I need them for projects and I don't have anything in the orthogonal world on my plate. If you want to take a stab at it pull requests are certainly always welcome! The data should be all there from the actual tmx file so the drawing is where most of the work would end up being (with some other minor additions like converting world to tile). It would probably be easiest to just subclass TiledTileLayer and make a TiledOrthogonalTileLayer so the code doesnt have to be split with if/elses for everything.

**Comment by Monix** [comment]
Just to clarify, Nez already supports orthogonal, and I'm interested in support for isometric (staggered is also not supported, but not of interest to me). And yep, I'm happy to contribute what I come up with back to Nez!

Agreed that subclassing would be the way to go as far as layers, but it does get a bit awkward with TiledMap, which also has some orthogonal logic in it such as worldToTilePositionX(). For now we can probably get away with conditions in the various helper functions, but I could see it becoming bloated in future.

**Comment by prime31** [comment]
Updated my comment to have the correct verbiage! Seeing as how orthogonal is far and away the most used it isn't too terrible that the TiledMap class contains those methods. Extension methods on TiledMap or other methods with names that are prefixed with "orthogonal" could solve that fairly cleanly.

**Comment by Monix** [comment]
I didn't even realise C# supported extensions, agreed that could be a nice compromise. Hopefully I have a nice pull request for you in future. :)

For now it might be worth formalising the orthogonal-only support in the documentation or something?

**Comment by meowgoocat** [comment]
I also am in need of isometric support from Tiled maps.  If you have made any headway, let me know, otherwise I will probably start working on a solution myself.

**Comment by Monix** [comment]
@meowgoocat I've made headway and have basically finished. Now that I know someone else needs it I'll try to get a pull request in the next few hours :)

**Comment by prime31** [comment]
Awesome! I will be travelling for a few days but I'll give this a look when I get back in town.

**Comment by Monix** [comment]
@prime31 Thanks!

@meowgoocat Feel free to clone my isometric branch if you can't wait for the merge.

**Comment by meowgoocat** [comment]
@Monix 
Thanks a lot.  This is a huge help.

**Comment by Vanit** [comment]
@meowgoocat Glad to help!

Also FYI everyone I've just changed my username, so don't be confused, I'm still me. :)

**Comment by Vanit** [comment]
Probably going to have a follow up pull request with a new component for depth sorted rendering of entities amongst the map tiles (ie rendering players infront/behind tall tiles). Pretty common usecase that isn't satisfied by the existing TiledMapComponent. Hopefully will have something to show in a few days.

**Comment by Vanit** [comment]
Hey @prime31, not sure if you're holding out for me based on my previous comment, but the current pull request will be final for a while (unless you think I should fix it). Basically I've since realised that vanilla Nez doesn't quite fit the game I want to make (isometric game with 3D physics), so I'll be working off my own fork of it that uses Vector3 etc (initially I wanted to just subclass what I needed to, but there's a lot of internal flags that make that difficult). Might be able to backport some useful things in the future, depending on how compatible it is.
